flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -906,11 +907,33 @@ private void resetAvailableTransactionalIdsPool(Collection<String> transactional
 	// ----------------------------------- Utilities --------------------------
 
 	private void abortTransactions(Set<String> transactionalIds) {
-		for (String transactionalId : transactionalIds) {
-			try (FlinkKafkaProducer<byte[], byte[]> kafkaProducer =
-					initTransactionalProducer(transactionalId, false)) {
-				// it suffice to call initTransactions - this will abort any lingering transactions
-				kafkaProducer.initTransactions();
+		// shortcut for non-exactly-once producers
+		if (transactionalIds.isEmpty()) {
+			return;
+		}
+
+		ForkJoinPool forkJoinPool = null;
+		try {
+			// limit the number of connections to the number that is used during runtime
```

Comment:
One more thought - what if kafka/zookeeper are limiting number of concurrent connections/operations? Technically speaking yes, previously we COULD have the same number of active `KafkaProducer`s during runtime but:
1. **usually**, only 2 of them were being used - here it's always `kafkaProducersPoolSize`. This argument maybe we can ignore, since we can argue that the user's setup should always support `kafkaProducersPoolSize` number of active `KafkaProducer`s.
2. however, previously `KafkaProducers` they were not being used concurrently. If we use them concurrently now, we can briefly flood kafka brokers. Can this be an issue? 

Can you @NicoK maybe clear this out with @becketqin? He has better understanding of how Kafka internally works.

Reply:
I would not concern too much about the load on the Kafka brokers. They are designed to accept thousands of requests per second. And committing / aborting the transactions in Kafka is pretty much just like send an ordinary message. So it is probably OK.

Reply:
I would not concern too much about the load on the Kafka brokers. They are designed to accept thousands of requests per second. And committing / aborting the transactions in Kafka is pretty much just like send an ordinary message. So it is probably OK.